### PR TITLE
Scenarios endpoints

### PR DESCRIPTION
**Pull request by Argn0**

### CODE REVIEW

File: routes/spec.js [hunk 1]
```diff
@@ -3943,6 +3944,186 @@ Please keep request rate to approximately 3/s.
         },
       },
     },
+    '/scenarios/{hero_id}/itemTimings': {
+      get: {
+        summary: 'GET /scenarios/{hero_id}/itemTimings',
+        description: 'Win rates for certain item timings',
```
Comment (howardchung) [inline routes/spec.js]: "on a hero".

You may want to consider flipping the param order to /itemTimings/hero_id if you want to support querying all item timings (not providing a hero_id) returns general item timings win rates

File: routes/spec.js [hunk 2]
```diff
@@ -3943,6 +3944,186 @@ Please keep request rate to approximately 3/s.
...
+        tags: ['scenarios'],
+        parameters: [{
+          name: 'item',
+          in: 'query',
+          description: 'Filter by item name',
```
Comment (howardchung) [inline routes/spec.js]: may want to provide an example item name

File: routes/spec.js (hunk 1 above)
Comment (howardchung) [reply routes/spec.js]: You may want to consider making both hero_id and item query params, so the user can filter on them independently.  e.g. a user can request win rates for blink dagger on all heroes, or win rates for all items on Riki

File: routes/spec.js [hunk 3]
```diff
@@ -3943,6 +3944,186 @@ Please keep request rate to approximately 3/s.
...
+                  item: {
+                    description: 'Purchased item',
+                    type: 'string',
+                  },
+                  time: {
+                    description: 'Ingame time in seconds before the item was purchased',
+                    type: 'integer',
+                  },
+                  games: {
+                    description: 'In how many games was the item bought by this hero within that time',
```
Comment (howardchung) [inline routes/spec.js]: "the number of games where the hero bought this item before this time"

Comment (howardchung) [reply routes/spec.js]: actually i forget, are the item timings rounded down to the nearest 5 minute block?

File: routes/spec.js [hunk 4]
```diff
@@ -3943,6 +3944,186 @@ Please keep request rate to approximately 3/s.
...
+                    type: 'string',
+                  },
+                  wins: {
+                    description: 'How many games were won when this hero bought the item within that time',
```
Comment (howardchung) [inline routes/spec.js]: the number of games won where the hero bought this item before this time

File: routes/spec.js [hunk 5]
```diff
@@ -3943,6 +3944,186 @@ Please keep request rate to approximately 3/s.
...
+                    description: 'Purchased item',
+                    type: 'string',
+                  },
+                  time: {
+                    description: 'Ingame time in seconds before the item was purchased',
+                    type: 'integer',
+                  },
+                  games: {
+                    description: 'In how many games was the item bought by this hero within that time',
+                    type: 'string',
```
Comment (howardchung) [inline routes/spec.js]: are these integers?

File: routes/spec.js [hunk 6]
```diff
@@ -3943,6 +3944,186 @@ Please keep request rate to approximately 3/s.
...
+                    description: 'How many games were won when this hero bought the item within that time',
+                    type: 'string',
+                  },
+                },
+              },
+            },
+          },
+        },
+        route: () => '/scenarios/:hero_id/itemTimings',
+        func: (req, res, cb) => {
```
Comment (howardchung) [inline routes/spec.js]: consider moving this function to queries.js, the eventual goal is to put all of them there (I haven't gotten around to some of them

File: routes/spec.js [hunk 7]
```diff
@@ -3943,6 +3944,186 @@ Please keep request rate to approximately 3/s.
...
+          const heroId = req.params.hero_id;
+          const item = req.query.item;
+
+          if (item) {
+            db.raw(`SELECT item, time, sum(games) games, sum(wins) wins
```
Comment (howardchung) [inline routes/spec.js]: you should be able to use a single query template for this by doing something like `AND (? IS NULL or item = ?)` (basically make the first condition true when no item/empty string/null is passed)

File: routes/spec.js [hunk 8]
```diff
@@ -3943,6 +3944,186 @@ Please keep request rate to approximately 3/s.
...
+                if (err) {
+                  return cb(err);
+                }
+                return res.json(result.rows);
+              });
+          }
+        },
+      },
+    },
+    '/scenarios/{hero_id}/laneRoles': {
```
Comment (howardchung) [inline routes/spec.js]: same comment about maybe inverting the order

File: routes/spec.js [hunk 9]
```diff
@@ -3943,6 +3944,186 @@ Please keep request rate to approximately 3/s.
...
+                    description: 'The hero\'s lane role',
+                    type: 'integer',
+                  },
+                  time: {
+                    description: 'Total up to game length in seconds',
+                    type: 'integer',
+                  },
+                  games: {
+                    description: 'In how many games this hero was played in this lane role',
+                    type: 'string',
```
Comment (howardchung) [inline routes/spec.js]: integer?

File: routes/spec.js [hunk 10]
```diff
@@ -3943,6 +3944,186 @@ Please keep request rate to approximately 3/s.
...
+              type: 'array',
+              items: {
+                type: 'object',
+                properties: {
+                  lane_role: {
+                    description: 'The hero\'s lane role',
+                    type: 'integer',
+                  },
+                  time: {
+                    description: 'Total up to game length in seconds',
```
Comment (howardchung) [inline routes/spec.js]: is this the full game length?

File: routes/spec.js [hunk 11]
```diff
@@ -3943,6 +3944,186 @@ Please keep request rate to approximately 3/s.
...
+                    type: 'integer',
+                  },
+                  games: {
+                    description: 'In how many games this hero was played in this lane role',
```
Comment (howardchung) [inline routes/spec.js]: same comment about phrasing on this scenario

File: util/scenariosUtil.js [hunk 12]
```diff
@@ -148,8 +148,15 @@ function validateMatchProperties(match) {
   return matchProperties.every(property => match[property] !== undefined && match[property] !== null);
 }
 
+const teamScenariosQueryParams = {
+  pos_chat_1min: 'Positivity in chat before 1min',
```
Comment (howardchung) [inline util/scenariosUtil.js]: expand this to "1 minute"

File: routes/spec.js (hunk 3 above)
Comment (Argn0) [reply routes/spec.js]: they are  [7.5, 10, 12, 15, 20, 25, 30]

so 0 - 7.5 min  ,  7.5 - 10 min , 10 - 12 min  etc.

File: routes/spec.js (hunk 5 above)
Comment (Argn0) [reply routes/spec.js]: no they are strings in the JSON response

**Comment by Argn0** [comment]
Moved all functions to queries.js and fixed everything else you mentioned.

File: routes/params.js [hunk 13]
```diff
@@ -201,4 +201,11 @@ module.exports = {
     required: false,
     type: 'integer',
   },
+  heroIdParamPath: {
+    name: 'hero_id',
+    in: 'path',
```
Comment (howardchung) [inline routes/params.js]: this is now a query param

File: routes/spec.js [hunk 14]
```diff
@@ -3943,6 +3944,178 @@ Please keep request rate to approximately 3/s.
...
+                  lane_role: {
+                    description: 'The hero\'s lane role',
+                    type: 'integer',
+                  },
+                  time: {
+                    description: 'Maximum game length in seconds',
+                    type: 'integer',
+                  },
+                  games: {
+                    description: 'The number of games won where the hero bought this item before this time',
```
Comment (howardchung) [inline routes/spec.js]: i think this the description for the field below,

This one should be just number of games

File: routes/spec.js [hunk 15]
```diff
@@ -3943,6 +3944,178 @@ Please keep request rate to approximately 3/s.
...
+                  item: {
+                    description: 'Purchased item',
+                    type: 'string',
+                  },
+                  time: {
+                    description: 'Ingame time in seconds before the item was purchased',
+                    type: 'integer',
+                  },
+                  games: {
+                    description: 'The number of games won where the hero bought this item before this time',
```
Comment (howardchung) [inline routes/spec.js]: this should be for the field below

File: store/queries.js [hunk 16]
```diff
@@ -1263,6 +1264,40 @@ function insertMatch(match, options, cb) {
   });
 }
 
+function getItemTimings(req, cb) {
+  const heroId = req.query.hero_id || 0;
+  const item = req.query.item || 0;
+  db.raw(
+    `SELECT hero_id, item, time, sum(games) games, sum(wins) wins FROM scenarios WHERE item IS NOT NULL
+       AND (0 = :heroId OR hero_id = :heroId) AND ('0' = :item OR item = :item)
```
Comment (howardchung) [inline store/queries.js]: does '0' comparison work?  shouldn't it be 0 since it is not a string?

Comment (howardchung) [reply store/queries.js]: Would advise splitting each WHERE/AND on its own line

File: store/queries.js [hunk 17]
```diff
@@ -1263,6 +1264,40 @@ function insertMatch(match, options, cb) {
...
+  )
+    .asCallback((err, result) => cb(err, result));
+}
+
+function getLaneRoles(req, cb) {
+  const heroId = req.query.hero_id || 0;
+  const lane = req.query.lane_role || 0;
+  db.raw(
+    `SELECT hero_id, lane_role, time, sum(games) games, sum(wins) wins 
+       FROM scenarios WHERE lane_role IS NOT NULL
```
Comment (howardchung) [inline store/queries.js]: same formatting of SQL

File: store/queries.js [hunk 18]
```diff
@@ -1263,6 +1264,40 @@ function insertMatch(match, options, cb) {
...
+    { heroId, lane },
+  )
+    .asCallback((err, result) => cb(err, result));
+}
+
+function getTeamScenarios(req, cb) {
+  const scenario = su.teamScenariosQueryParams[req.query.scenario] || 0;
+
+  db.raw(`SELECT scenario, is_radiant, region, sum(games) games, sum(wins) wins 
+    from team_scenarios WHERE (0 = :scenario OR scenario = :scenario)
```
Comment (howardchung) [inline store/queries.js]: format the query

File: util/scenariosUtil.js [hunk 19]
```diff
@@ -121,16 +122,16 @@ const scenarioChecks = {
           }
         }
         if (radiantNegative) {
-          rows.push(buildTeamScenario('Negativity in chat before 1min', true, match)[0]);
+          rows.push(buildTeamScenario('Negativity in chat before 1 minute', true, match)[0]);
```
Comment (howardchung) [inline util/scenariosUtil.js]: you can probably reuse the teamScenariosQueryParams strings here

**Comment by Argn0** [comment]
- formatted all query strings
- fixed games/wins response descriptions
- removed heroIdParamPath from params.js I was not using it anymore

File: routes/spec.js [hunk 20]
```diff
@@ -3943,6 +3944,178 @@ Please keep request rate to approximately 3/s.
...
+              type: 'array',
+              items: {
+                type: 'object',
+                properties: {
+                  scenario: {
+                    description: 'The scenario\'s name or description',
+                    type: 'string',
+                  },
+                  is_radiant: {
+                    description: 'Radiant or Dire',
```
Comment (howardchung) [inline routes/spec.js]: "boolean indicating if the player was radiant" (i feel like we have this property a few times already, it should probably be refactored into one place)

File: routes/spec.js [hunk 21]
```diff
@@ -3943,6 +3944,178 @@ Please keep request rate to approximately 3/s.
...
+                    type: 'boolean',
+                  },
+                  region: {
+                    decription: 'Region the game was played in',
+                    type: 'integer',
+                  },
+                  games: {
+                    description: 'The number of games where this scenario occured',
```
Comment (howardchung) [inline routes/spec.js]: occurred

File: store/queries.js [hunk 22]
```diff
@@ -1263,6 +1264,48 @@ function insertMatch(match, options, cb) {
...
+  db.raw(
+    `SELECT hero_id, item, time, sum(games) games, sum(wins) wins
+     FROM scenarios
+     WHERE item IS NOT NULL
+     AND (0 = :heroId OR hero_id = :heroId)
+     AND ('' = :item OR item = :item)
+     GROUP BY hero_id, item, time ORDER BY time, hero_id, item
+     LIMIT 1600`,
+    { heroId, item },
+  ).asCallback((err, result) => cb(err, result));
```
Comment (howardchung) [inline store/queries.js]: you can actually shorthand this to `asCallback(cb)`, but this is fine too, it makes it easier to add logging if we need to

File: routes/spec.js [hunk 23]
```diff
@@ -238,7 +239,7 @@ Please keep request rate to approximately 3/s.
                   type: 'integer',
                 },
                 radiant_win: {
-                  description: 'Boolean indicating whether Radiant won the match',
+                  description: radiantWinDescr,
```
Comment (howardchung) [inline routes/spec.js]: probably the entire radiant_win object can be refactored out

Comment (howardchung) [reply routes/spec.js]: you might actually want to create a new file in https://github.com/odota/core/tree/master/routes called properties.js and put commonly used properties in there, then import them into this file

File: routes/spec.js [hunk 24]
```diff
@@ -4085,15 +4086,15 @@ Please keep request rate to approximately 3/s.
                     type: 'string',
                   },
                   is_radiant: {
-                    description: 'Radiant or Dire',
+                    description: radiantWinDescr,
```
Comment (howardchung) [inline routes/spec.js]: this one is not the same as the others

**Pull request merged**